lib/dynamoid/fields.rb

```diff
@@ -303,12 +304,12 @@ def write_attribute(name, value)
         association.reset
       end
 
-      attribute_will_change!(name) # Dirty API
-
       @attributes_before_type_cast[name] = value
 
       value_casted = TypeCasting.cast_field(value, self.class.attributes[name])
       attributes[name] = value_casted
+
+      attribute_will_change!(name) if old_value != value_casted # Dirty API
```

Comment:
`attribute_will_change!` should be called before assigning a new value because it saves the previous old one.

Reply:
I fixed it!